samza-core/src/main/java/org/apache/samza/operators/stream/OutputStreamInternal.java

```diff
@@ -17,38 +17,28 @@
  * under the License.
  */
 
-package org.apache.samza.operators.data;
+package org.apache.samza.operators.stream;
 
 import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.operators.MessageStream;
+import org.apache.samza.operators.OutputStream;
+import org.apache.samza.system.StreamSpec;
+
+import java.util.function.Function;
 
 
 /**
- * An entry in the input/output {@link org.apache.samza.operators.MessageStream}s
+ * Internal representation of an output stream.
+ *
+ * @param <M> the type of messages in the output stream
  */
 @InterfaceStability.Unstable
-public interface MessageEnvelope<K, M> {
-
-  /**
-   * Get the key for this {@link MessageEnvelope}.
-   *
-   * @return  the key for this {@link MessageEnvelope}
-   */
-  K getKey();
-
-  /**
-   * Get the message in this {@link MessageEnvelope}.
-   *
-   * @return  the message in this {@link MessageEnvelope}
-   */
-  M getMessage();
-
-  /**
-   * Whether this {@link MessageEnvelope} indicates deletion of a previous message with this key.
-   *
-   * @return  true if the current {@link MessageEnvelope} indicates deletion of a previous message with this key
-   */
-  default boolean isDelete() {
-    return false;
-  }
+public interface OutputStreamInternal<K, V, M> extends MessageStream<M>, OutputStream<K, V, M> {
```

Comment:
Because `IntermediateStreamInternal implements OutputStreamInternal, InputStreamInternal`, but can't extend the two concrete classes directly.